### PR TITLE
feat(device-selection): fall back to using label for preferred device

### PR DESCRIPTION
**Pull request by virtuacoplenny**
Prefer to use the stored device id. If it cannot be found, but
a similarly matching label is found, use the device id from the
device with the similarly matching label. This is intended
to solve the case where the the device has changed USB ports.

### CODE REVIEW

**Comment by virtuacoplenny** [comment]
https://github.com/jitsi/jitsi-meet/pull/3226 should go in first and then this PR re-tested. 3226 changes timing a bit to make device switch on device list change more likely to pick up the preferred device that is plugged in.

**Comment by virtuacoplenny** [comment]
Force pushed to resolve conflicts are JitsiMeetJS.mediaDevices.isDeviceListAvailable() no longer returning a promise.

**Comment by bbaldino** [comment]
Do we have a sense for the risk of an unexpected device being selected due to the loose label matching?

**Comment by virtuacoplenny** [comment]
I recall believing it was low risk and I'm usually right or wrong or a mixture of both. The loose matching tries to eliminate what operating systems append to the usb devices; the matching can be found in the regex used in the functions.js file. The idea is this could be a problem if there are multiples of the same camera/mic plugged in and ports are switched but for most cases use cases targeting removal of a specific subset of the label will not be harmful.

**Comment by virtuacoplenny** [comment]
Force pushed to resolve conflicts introduced with removal of state from mediaDeviceHelper. The conflicts were in conference#_initDeviceList.

File: react/features/base/settings/functions.js [hunk 1]
```diff
@@ -5,6 +5,119 @@ import { toState } from '../redux';
...
+    // Prioritize matching the deviceId of the camera.
+    if (foundMatchingBasedonDeviceId) {
+        return cameraDeviceId;
+    }
+
+    // Fall back to doing a partial match of the label if a matching deviceId
+    // has not been found. Operating systems may append " #{number}" somewhere
+    // in the label so find and strip that bit.
+    const lastSpacePoundNumRegex = /\s#\d*(?!.*\s#\d*)/;
+    const replacement = '';
```
Comment (bbaldino) [inline react/features/base/settings/functions.js]: i'd name this `empty` or `nothing`?

File: react/features/base/settings/functions.js [hunk 2]
```diff
@@ -5,6 +5,119 @@ import { toState } from '../redux';
...
+
+/**
+ * Searches known devices for a matching deviceId and fall back to matching on
+ * label. Returns the stored preferred micDeviceId if a match is not found.
+ *
+ * @param {Object|Function} stateful - The redux state object or
+ * {@code getState} function.
+ * @returns {string}
+ */
+export function getPreferredMicDeviceId(stateful: Object | Function) {
```
Comment (bbaldino) [inline react/features/base/settings/functions.js]: is there not a way to have a helper function which does the common logic (label fallback, stripping of the os suffix, etc.) that both audio and video device logic could use?

Comment (virtuacoplenny) [reply react/features/base/settings/functions.js]: I couldn't think of it without passing in several options, at which point I wondered if what worthwhile. For example, I look at different keys in features/base/settings, I look at a different lists of devices in features/base/device, I use a different regex,  and I use a different replacement character. I thought of breaking it up into little helpers too but thought it didn't really make things better. I can keep thinking about it now that I suspect device issues were due to lack of persistence and not the label changing.

File: conference.js [hunk 3]
```diff
@@ -731,7 +739,14 @@ export default {
                     this.setVideoMuteStatus(true);
                 }
 
-                this._initDeviceList();
+                // Initialize device list a second time to ensure device labels
+                // get populated in case of an initial gUM acceptance; otherwise
+                // they may remain as empty strings.
```
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply conference.js]: Correct. On initial gum, without user permissions, enumerateDevices might not return labels. If a/v is accepted, then enumerateDevices returns labels.

So the idea here is if this is a first time user the labels here will be empty. In which case the second enumerate devices should be called to make sure the labels are available for the app (and label matching isn't necessary anyway). If the user is returning, then we need to call enumerateDevices early to get labels and do the matching of new label to device id.

File: conference.js [hunk 4]
```diff
@@ -2085,12 +2100,18 @@ export default {
...
                 })
-                .then(() => {
+                .then(stream => {
                     logger.log('switched local video device');
+
+                    const videoTrack = stream && stream.getTrack();
+
                     APP.store.dispatch(updateSettings({
-                        cameraDeviceId
+                        cameraDeviceId,
```
[the remark this answers is not in the document]
Comment (virtuacoplenny) [reply conference.js]: I would argue the whole chain probably shouldn't be called if null/undefined. I think the UI does not have a flow right now (outside of a UI bug that probably by default has a non-zero chance of happening) where null or undefined can be passed in when the user has selected a new video device to use in the settings modal.

File: react/features/base/settings/functions.js (hunk 2 above)
Comment (virtuacoplenny) [reply react/features/base/settings/functions.js]: I've now gone with the approach of passing in options.

Comment (virtuacoplenny) [reply react/features/base/settings/functions.js]: For the first time I can remember I have blown away changes I wanted to keep by doing a force push; force pushing has finally bit me hard. I redid the changes to abstract the logic into one place.

**Pull request closed without merging**